File: 2022/generateParenthesis.js
let generateParenthesis = (n) => {
    const parens = [];
    recur(n, 1, 0, '(', parens);
    return parens;
}

let recur = (n, l, r, str, parens) => {
    if (str.length >= n*2) {
        parens.push(str);
        return;
    };
    
    if (l < n) {
        recur(n, l+1, r, str + '(', parens);
    }

    if (r < l) {
        recur(n, l, r+1, str + ')', parens);
    };
};

File: 2022/generate_parenthesis.py
class Solution(object):

    parenthesis = []

    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self.recurse(n, 1, 0, '(')
        return self.parenthesis

    def recurse(self, n, l, r, str):
        if len(str) == n*2:
            self.parenthesis.append(str)
            
        if (l<n):
            self.recurse(n, l+1, r, str + '(')
            
        if (r<l):
            self.recurse(n, l, r+1, str + ')')